src/main/java/com/google/api/generator/engine/ast/ThisObjectValue.java

```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+
+@AutoValue
+public abstract class ThisObjectValue implements ObjectValue {
+  private static final String THIS_VALUE = "this";
+
+  public abstract TypeNode type();
+
+  @Override
+  public String value() {
+    return THIS_VALUE;
+  }
+
+  public static ThisObjectValue withType(TypeNode type) {
+    return builder().setType(type).build();
+  }
+
+  private static Builder builder() {
+    return new AutoValue_ThisObjectValue.Builder();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setType(TypeNode type);
+
+    public abstract ThisObjectValue autoBuild();
+
+    public ThisObjectValue build() {
+      ThisObjectValue thisObjectValue = autoBuild();
+      Preconditions.checkState(
+          TypeNode.isReferenceType(thisObjectValue.type()), "this can only refer to object types");
+      Preconditions.checkState(
```

Comment:
I add this new check here to make sure the `this` type is belongs to custom class object. I see all the case of using this is inside of class definition, like `public Class Student {}`. This check is to exclude java build-in class like `java.lang.*`. [example](https://github.com/googleapis/gapic-generator-java/pull/137/files#diff-eb673d6d6c1d0400da66863c9b86eddfR39) [implementation](https://github.com/googleapis/gapic-generator-java/pull/137/files#diff-3eb77b740c6a645ccd1ef7aa836d7348R108)

But I am not pretty sure about this. Open to your advice and suggestion @miraleung @xiaozhenliu-gg5. Thanks a lot.

Reply:
I think we don't need this check because `this` can be misused in many non-`java.lang` classes, which span all Java libs or other deps. Since checking the entire space is not feasible, we may as well not have this check.

Reply:
Sounds reasonable. removed. Thanks.